rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -0,0 +1,119 @@
+---
+authors: Alex McGrath (alex.mcgrath@goteleport.com
+state: draft
+---
+
+# RFD 57 - Automatic discovery and enrollment of AWS servers
+
+## What
+
+Proposes a way by which an auth service might automatically discover and
+register AWS EC2 instances.
+
+## Why
+
+Currently when adding a new AWS server, its required that Teleport be installed
+after the server has been provisioned which may be a slow process for
+organizations with large numbers of servers as it needs to be installed and then
+added to the teleport cluster
+
+With the changes described in this document, Teleport will be able to resolve
+the issues with adding AWS servers to Teleport clusters automatically.
+
+
+## Discovery
+
+A Teleport auth agent will need to be configured with an assume role policy for
+each of the accounts that the agent will discover services on as described by
+the EC2 join method docs[1]
+
+Discovery can use a matcher similar to the `db_service/aws` matcher:
+
+```yaml
+auth_service:
+  enabled: "yes"
+  aws:
+  - types: ["ec2"]
+    regions: ["us-west-1"]
+    tags:
+      "teleport": "yes"
+    accounts:
+    - aws_account: "222222222222"
+      aws_roles:
+      - "arn:aws:iam::222222222222:role/teleport-DescribeInstances-role"
+      - "arn:aws:iam::222222222222:role/teleport-Install-role"
+```
+
+The agent will use EC2's `DescribeInstances` API in order to list instances[1]. This
+will require the teleport auth agent to include `ec2:DescribeInstances` as part of
+it's IAM permissions
+
+As with database AWS discovery, new EC2 nodes will discovered periodically on a 5
+minute timer, and any new nodes will be added and existing instances updated/deleted.
+
+The EC2 join method already available in Teleport will be used to add nodes to the
+Teleport cluster, this sets the node name to the EC2 `<account-id>-<instance-id>`
+which can be reused here to avoid attempting to reinstall teleport on an instance
+where it is already present.
```

Comment:
Instead of the node name, can we use labels or instance metadata to mark instances that we have already installed? What happens if the instance was installed, but stopped the heartbeat?